index.js

```diff
@@ -75,19 +84,27 @@ class Main {
 
   async workLoop() {
     while (this.shouldWork) {
+      if (this.taskManager.queue.size < constantsBase.PQUEUE_MAX_SIZE) {
+        const interval = await nextIntervalCalculator(this.worker);//TODO:
+        if (interval.success) {
+          delete interval.success;
+          this.taskManager.pushToQueue(() => this.worker.work(interval));
+          this.worker.lastExportedBlock = interval.toBlock;
+        }
+      }
       this.worker.lastRequestStartTime = new Date();
-      const events = await this.worker.work();
-
       this.worker.lastExportTime = Date.now();
 
       this.updateMetrics();
       this.lastProcessedPosition = this.worker.getLastProcessedPosition();
 
-      if (events && events.length > 0) {
-        await this.exporter.storeEvents(events);
+      if (this.taskManager.buffer && this.taskManager.buffer.length > 0) {
```

Comment:
Currently the consumer of the buffer is part of the main loop. I think it would make sense to separate it. Make the buffer blocking (wrap it into a simple class where you can call `await waitOnData()`) and then start the consumer and make it block on the buffer.